MINOR VerifableProducer ducktape can set idempotency and retries

josefk31 opened the pull request:
It is currently impossible to set both number of retries and idempotency
in the DucktapeVerifiable producer. This change allows that to occur.

josefk31 committed:
MINOR VerifableProducer ducktape can set idempotency and retries

It is currently impossible to set both number of retries and idempotency in the DucktapeVerifiable producer. This change allows that to occur.

josefk31 commented:
Note; there are no current examples where both of these `enable_idempotent` and `retries` are set at the same time. So this change won't affect existing tests. 

I'm adding this because I am debugging some flaky test failures and want these to alter the configurations. 

Testing:

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-04-03--003
run time:         4 minutes 18.534 seconds
tests run:        2
passed:           2
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.eligible_leader_replicas_test.EligibleLeaderReplicasTest.test_basic_eligible_leader_replicas.metadata_quorum=ISOLATED_KRAFT.group_protocol=classic
status:     PASS
run time:   2 minutes 9.873 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.eligible_leader_replicas_test.EligibleLeaderReplicasTest.test_basic_eligible_leader_replicas.metadata_quorum=ISOLATED_KRAFT.group_protocol=consumer
status:     PASS
run time:   2 minutes 8.346 seconds
--------------------------------------------------------------------------------
```

FrankYang0529 commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -147,9 +147,10 @@ def _worker(self, idx, node):
-        elif self.retries is not None:
+            if self.retries is None:
+                producer_prop_file += "\nretries=1000000\n"
+        if self.retries is not None:
```
In retries document https://kafka.apache.org/documentation/#producerconfigs_retries, it mentions
```
Enabling idempotence requires this config value to be greater than 0.
If conflicting configurations are set and idempotence is not explicitly enabled, idempotence is disabled.
```

I think we should add assertion to `__int__` to make sure if enable_idempotence is true, then retries should not be zero.

FrankYang0529 reviewed (commented):
Thanks for the patch. Leave a minor comment.

josefk31 committed:
Added assertion

When retries and idempotence are enabled, make sure retries are a
positive value.

josefk31 commented:
@FrankYang0529 Thanks for review - I've updated the code with suggestion. Tests are still green. PTAL.

josefk31 committed:
Undid unneeded change

FrankYang0529 reviewed (approved):
LGTM. Thanks.
